packages/flutter_tools/test/general.shard/terminal_handler_test.dart

```diff
@@ -4,225 +4,172 @@
-      expect(testRunner.hasHelpBeenPrinted, true);
-    });
+    expect(testRunner.hasHelpBeenPrinted, false);
+    await terminalHandler.processTerminalInput('h');
+    expect(testRunner.hasHelpBeenPrinted, true);
+  });
 
-    testUsingContext('help character surrounded with newlines', () async {
-      final TestRunner testRunner = createTestRunner();
-      final TerminalHandler terminalHandler = TerminalHandler(testRunner);
-      expect(testRunner.hasHelpBeenPrinted, false);
-      await terminalHandler.processTerminalInput('\nh\n');
-      expect(testRunner.hasHelpBeenPrinted, true);
-    });
+  testWithoutContext('keyboard input handling help character surrounded with newlines', () async {
+    final TestRunner testRunner = TestRunner();
+    final Logger logger = BufferLogger.test();
+    final Signals signals = Signals.test();
+    final Terminal terminal = Terminal.test();
+    final TerminalHandler terminalHandler = TerminalHandler(
+      testRunner,
+      logger: logger,
+      signals: signals,
+      terminal: terminal,
+    );
+
+    expect(testRunner.hasHelpBeenPrinted, false);
+    await terminalHandler.processTerminalInput('\nh\n');
+    expect(testRunner.hasHelpBeenPrinted, true);
   });
 
   group('keycode verification, brought to you by the letter', () {
     MockResidentRunner mockResidentRunner;
     TerminalHandler terminalHandler;
+    BufferLogger testLogger;
 
     setUp(() {
+      testLogger = BufferLogger.test();
+      final Signals signals = Signals.test();
+      final Terminal terminal = Terminal.test();
       mockResidentRunner = MockResidentRunner();
-      terminalHandler = TerminalHandler(mockResidentRunner);
+      terminalHandler = TerminalHandler(
+        mockResidentRunner,
+        logger: testLogger,
+        signals: signals,
+        terminal: terminal,
+      );
       when(mockResidentRunner.supportsServiceProtocol).thenReturn(true);
     });
 
-    testUsingContext('a, can handle trailing newlines', () async {
+    testWithoutContext('a, can handle trailing newlines', () async {
       await terminalHandler.processTerminalInput('a\n');
 
       expect(terminalHandler.lastReceivedCommand, 'a');
     });
 
-    testUsingContext('n, can handle trailing only newlines', () async {
+    testWithoutContext('n, can handle trailing only newlines', () async {
       await terminalHandler.processTerminalInput('\n\n');
 
       expect(terminalHandler.lastReceivedCommand, '');
     });
 
-    testUsingContext('a - debugToggleProfileWidgetBuilds with service protocol', () async {
+    testWithoutContext('a - debugToggleProfileWidgetBuilds with service protocol', () async {
       await terminalHandler.processTerminalInput('a');
 
       verify(mockResidentRunner.debugToggleProfileWidgetBuilds()).called(1);
     });
 
-    testUsingContext('a - debugToggleProfileWidgetBuilds without service protocol', () async {
-      when(mockResidentRunner.supportsServiceProtocol).thenReturn(false);
```

Comment:
Shame we lost the negative case logic.  Can you add it back to https://github.com/flutter/flutter/blob/master/packages/flutter_tools/test/general.shard/resident_runner_test.dart#L1653?

Reply:
yeah - I need to take another pass on this and see where we lost test coverage (I think some of these are covered, but definitely not all of them).